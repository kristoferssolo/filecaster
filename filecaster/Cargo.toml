[package]
name = "filecaster"
version = "0.2.3"
edition = "2024"
authors = ["Kristofers Solo <dev@kristofers.xyz>"]
description = "Procedural macro to derive configuration from files, with optional merging capabilities."
license = "MIT OR Apache-2.0"
repository = "https://github.com/kristoferssolo/filecaster"
homepage = "https://github.com/kristoferssolo/filecaster"
documentation = "https://docs.rs/filecaster"
readme = "../README.md"
keywords = ["configuration", "file-parsing"]
categories = ["rust-patterns", "parsing", "config"]
exclude = ["/.github", "/.gitignore", "/tests", "*.png", "*.md"]

[features]
default = ["serde", "derive"]
derive = ["dep:filecaster-derive"]
serde = ["dep:serde", "filecaster-derive/serde"]
merge = ["dep:merge", "filecaster-derive/merge"]

[dependencies]
filecaster-derive = { workspace = true, optional = true }
serde = { workspace = true, optional = true }
merge = { workspace = true, optional = true }

[dev-dependencies]
serde_json.workspace = true
tempfile.workspace = true
toml.workspace = true
trybuild.workspace = true
