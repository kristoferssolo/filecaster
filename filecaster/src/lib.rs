pub use filecaster_derive::FromFile;
use serde::{Deserialize, Serialize};

/// Marker for types that can be built from an `Option<Shadow>` produced by the macro.
pub trait FromFile: Sized {
    fn from_file(file: Option<Self::Shadow>) -> Self;

    /// Associated shadow type generated by the macro.
    type Shadow: Default + Serialize + for<'de> Deserialize<'de>;
}

impl<T> FromFile for T
where
    T: Default + Serialize + for<'de> Deserialize<'de>,
{
    type Shadow = T;
    fn from_file(file: Option<Self::Shadow>) -> Self {
        file.unwrap_or_default()
    }
}
