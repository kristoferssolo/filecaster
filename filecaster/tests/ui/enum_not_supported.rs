use filecaster::FromFile;

#[derive(FromFile)]
enum MyEnum {
    A,
    B,
}

fn main() {}
