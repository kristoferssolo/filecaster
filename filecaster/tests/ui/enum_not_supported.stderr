error: FromFile only works on structs.
       Enums are not supported.
 --> tests/ui/enum_not_supported.rs:4:6
  |
4 | enum MyEnum {
  |      ^^^^^^
